sheeprl/algos/dreamer_v1/dreamer_v1.py

```diff
@@ -645,36 +645,40 @@ def main(fabric: Fabric, cfg: Dict[str, Any]):
 
         # Train the agent
         if iter_num >= learning_starts:
-            ratio_steps = policy_step - prefill_steps * policy_steps_per_iter
-            per_rank_gradient_steps = ratio(ratio_steps / world_size)
-            if per_rank_gradient_steps > 0:
-                with timer("Time/train_time", SumMetric, sync_on_compute=cfg.metric.sync_on_compute):
-                    sample = rb.sample_tensors(
-                        batch_size=cfg.algo.per_rank_batch_size,
-                        sequence_length=cfg.algo.per_rank_sequence_length,
-                        n_samples=per_rank_gradient_steps,
-                        dtype=None,
-                        device=device,
-                        from_numpy=cfg.buffer.from_numpy,
-                    )  # [N_samples, Seq_len, Batch_size, ...]
-                    for i in range(per_rank_gradient_steps):
-                        batch = {k: v[i].float() for k, v in sample.items()}
-                        train(
-                            fabric,
-                            world_model,
-                            actor,
-                            critic,
-                            world_optimizer,
-                            actor_optimizer,
-                            critic_optimizer,
-                            batch,
-                            aggregator,
-                            cfg,
-                        )
-                        cumulative_per_rank_gradient_steps += 1
-                    train_step += world_size
-                if aggregator:
-                    aggregator.update("Params/exploration_amount", actor._get_expl_amount(policy_step))
+            is_distributed = fabric.world_size > 1
+            if (
+                cfg.algo.train_on_episode_end and reset_envs > 0 and not is_distributed
```

Comment:
Hi @LucaVendruscolo, here I see a problem: if you set `cfg.algo.train_on_episode_end = True` and you start a distributed training, then you will hav e the following situation:
- `cfg.algo.train_on_episode_end = True`
- `reset_envs > 0 = True` (let us suppose that the episode ended)
- `not is_distribured = False`
- `not cfg.algo.train_on_episode_end = False`

This becomes: `(True and True and False) or False = False`

In this case, the agent will never enter in the if statement, so the agent will never be trained.
What is missing is the modification of the config `cfg.algo.train_on_episode_end` when `is_distributed` is `True`.
For example, by adding near row 385 something like this:
```python
if fabric.world_size > 1:
    cfg.algo.train_on_episode_end = False
```

Or you need to modify the condition in order to take into account the situation described above